### ISSUE TITLE
Why 3 classes instead of 2?

### ISSUE THREAD

**Issue by voldemortX** [liuruijin17/LSTR#43]
@liuruijin17 Hi! I have a small question: would you kindly tell me why use 3 classes? Labels are all 1(lane)/2(no object), it seems 2 classes should suffice. Then there is another question, maybe one can directly use BCE loss here with only 1-dim output with [`class_embed`](https://github.com/liuruijin17/LSTR/blob/6044f7b2c5892dba7201c273ee632b4962350223/models/py_utils/kp.py#L162)?
I've checked the DETR repo for discussions on single class detection and still can't figure out why you use 3-dim output at `class_embed`.

EDIT:
Is it related to COCO index starts at 1?

**Comment by liuruijin17** [comment]
Hi, a great question. 

I think it's a bug.

When I first wrote the code I just followed DETR to add an empty class, since authors of DETR said they set 91 classes, not 90 on COCO because [ "...  don't need to carry a class mapping, ..."](https://github.com/facebookresearch/detr/issues/23). 

However, this makes no sense in LSTR and does not hold in our task.

Besides, today one of my colleagues also asked me this question and said she used 2 and it also worked.

As for "BCE loss with only 1-dim output", I think this should also work. 

Appreciate your questions! Thanks a lot!

**Comment by voldemortX** [comment]
@liuruijin17 Thanks for the answer! One more question, does your colleague experience a slight performance diff between 3 classes and 2? I think maybe this could impact the best hyperparameter for loss weightings, since in my re-implementation, the train-set-only accuracy seems ~0.5 percent lower than when trained on your original code. Although there may be other details I did not think of...

**Comment by liuruijin17** [comment]
https://github.com/liuruijin17/LSTR/blob/6044f7b2c5892dba7201c273ee632b4962350223/models/py_utils/detr_loss.py#L46
She changes the second argument to be zero since the above line would initialize target_classes as a full '2' tensor ('2' for background)

https://github.com/liuruijin17/LSTR/blob/6044f7b2c5892dba7201c273ee632b4962350223/models/py_utils/detr_loss.py#L33
then changes the empty_weight is initialized by `torch.ones(self.num_classes)`

The difference is
(1) 3 classes output, '1': lane, '2' background, '0' no supervision
(2) 2 classes output, '1': lane, '0' background

She trained on another dataset, so I would train this version when our GPUs have free spaces. If the accuracy also drops, then there still may have bugs that should be fixed.

**Comment by voldemortX** [comment]
Thanks! I'll also try some hyper parameter tuning on my code.

**Comment by liuruijin17** [comment]
Hi, @voldemortX 
I trained the 2 classes output version which achieves 96.22 Accuracy on the TuSimple test set.
![cls2results](https://user-images.githubusercontent.com/29913368/120980064-cf393c80-c7a8-11eb-898d-b9a861ccb0d9.JPG)

The trained model can be downloaded at:
https://drive.google.com/file/d/1mJCGRbBKARQutsRoOVgyeixLp9oD5kYY/view?usp=sharing

**Comment by liuruijin17** [comment]
Hi, @voldemortX 
I trained the 2 classes output version which achieves 96.22 Accuracy on the TuSimple test set.
![cls2results](https://user-images.githubusercontent.com/29913368/120980064-cf393c80-c7a8-11eb-898d-b9a861ccb0d9.JPG)

The trained model can be downloaded at:
https://drive.google.com/file/d/1mJCGRbBKARQutsRoOVgyeixLp9oD5kYY/view?usp=sharing

**Comment by voldemortX** [comment]
@liuruijin17 Thanks! So 2/3 classes should not affect best hyper-parameters. Maybe my slightly lower performance comes from not freezing the first BN layer.
